### ISSUE TITLE
Spaced repetition system

### ISSUE THREAD

**Issue by hwgilbert16** [hwgilbert16/scholarsome#11]
SRS is a large part of many other flashcard software, such as Anki. Scholarsome’s implementation of an SRS will be more user-friendly while still being just as powerful.

**Comment by hwgilbert16** [comment]
See comment for info on the status of SRS https://github.com/hwgilbert16/scholarsome/pull/58#issuecomment-1806613996

**Comment by otfried** [comment]
I have used a number of different flash card apps with spaced repetition, and for me the one offered by memrise works best - unfortunately, they are no longer interested in user-made or community decks...

I never got used to Anki - it requires too much effort to classify cards into very easy - easy - hard - very hard, etc., and the repetition interval grows too fast - very soon I had intervals of two years or so :-(  

My suggestion would be to stick with "Know" and "Don't know" when testing words.   Memrise stores the time a card was last tested,  the total number of times it was tested,  how many times the user answered correctly, the current repetition interval (which seems to have been tuned quite nicely), and whether the word is "difficult" (this seems to be based on the number of "Don't know" answers, perhaps as a percentage of the number of tests.)   You can then later review the difficult words separately, and in smaller groups.

**Comment by hwgilbert16** [comment]
> I never got used to Anki - it requires too much effort to classify cards into very easy - easy - hard - very hard, etc., and the repetition interval grows too fast

Yes, this has been something I have been wrestling with and why SRS has been in the works for over 6 months.

The issue you mention with Anki is the exact reason why I haven't implemented their method of SRS. When you're working on scales of thousands of cards, having to measure your response between 5 different options for *every card* requires a lot of thinking and is inefficient.

> Memrise stores the time a card was last tested, the total number of times it was tested, how many times the user answered correctly, the current repetition interval

Sadly, this isn't a perfect method either. Memrise's method of repetition is based on machine learning. Realistically, this isn't practical to implement within Scholarsome. The reason it's so effective is because it collects many different factors and all uses them to gauge how well content is being learned. It's unrealistic to expect people who self-host to dedicate a significant amount of computational power to train a model, considering it would have to be done for every single deck and for every single user.

When an SRS does eventually get implemented, it will likely be reminiscent of Anki but with some tweaking and with significant user customizability. I'm thinking the [Leitner system](https://en.wikipedia.org/wiki/Leitner_system) will be the best one to implement.

In the future, it's possible an ML-based algorithm will be developed and offered as opt-in, but I don't foresee this happening for a long while.

**Comment by otfried** [comment]
I have used Memrise for 10 years, and never noticed it using machine learning.  Do they claim they do?  I suspect that's a marketing gimmick - there have never been any changes in the intervals, independent of whether I was active or not, and they are the same in all of my (many) decks.

Their algorithm is actually quite simple - one figures out the intervals they are using after a while, and one can actually verify them by looking at the network traffic.

For each card, Memrise stores the following attributes:
1. creation time 
2. time the card was last tested
3. time the card is due to be tested
4. total number of times the card was tested (incremented every time, even within one session)
5. total number of times the answer was correct in a test
6. streak length (incremented for every correct test, reset on incorrect test)
7. the current interval 

The intervals they use are:
4h 12h 1d 6d 12d 24d 48d 96d 180d

A new card gets a streak of 0 and an interval of 4h.  On each correct test, the streak increases by one, the interval is increased to the next one, but 180d is the last value - it does not increase from there.

On an incorrect test, the interval is set to 12h (unless it already was 12h, then it is set to 4h).  The streak length is then set to **-1**.   The card will be retested later in the same session, until it is answered correctly - then the streak will be set to 1 again.  (So a test session does not end until all cards in the test have been answered correctly.  If one aborts a test session, then the -1 streak length stays on the card and indicates that it needs to be tested immediately again in the next test session.)

Memrise adds some additional bells and whistles:  For instance, it has the concept of "nearly correct test" - when the edit distance between correct answer and the entered answer is small enough, it accepts that as nearly correct.  In that case, the streak is reset, a new test is scheduled for 4h later, but you actually keep your current interval - that's why the difference between last tested and due to be tested is not always the interval.  Memrise also has an algorithm to detect "difficult" cards.    I think both "nearly correct" and "difficult" are not really important.

The Memrise intervals work quite well for me.  Notice the big gap between 1d and 6d.   But of course one could actually make the sequence of interval times configurable for the user.

I have used something like the Leitner system when I studied with paper cards many years ago.   Sebastian Leitner did not have a computer and electronic cards, so he designed a system that achieves the effect of spaced learning using paper cards and boxes.   Today, with the ability to store attributes for our cards, we can simply store the due time for each card, and when creating a test session, we query the database for all cards whose due time has passed.  That also works when users do not practice in regular intervals (which is the assumption underlying Leitner's implementation).

**Comment by hwgilbert16** [comment]
> Do they claim they do? I suspect that's a marketing gimmick - there have never been any changes in the intervals, independent of whether I was active or not, and they are the same in all of my (many) decks.

I must have misread their algorithm basis - you're right, it seems like they do not integrate ML-features.

The method you describe for Memrise seems promising. Looks like it fixes many of Anki's shortfalls and adds some more on top of it. I'll definitely take this into account when I restart work on the SRS implementation. Thank you for the information.

**Comment by gabbard** [comment]
I'd recommend looking at the FSRS algorithm used by recent versions of Anki, which is pretty simple to implement and is almost universally considered greatly superior to their older algorithm (in my own experience it greatly lowered my review burden while maintaining retention rates): https://github.com/open-spaced-repetition/fsrs4anki/wiki/The-Algorithm

**Comment by A-Dalton** [comment]
As a daily user of Scholarsome, I recently forked the repo to continue working on and updating the software. So far, I've added dark mode, updated most of the tech stack, and implemented a very crude "Previous mistakes" feature that simply tracks wrong answers. Now I want to continue the gradual updates by implementing a spaced repetition system.

First off, thanks for the helpful input in this thread! In case anyone is interested or would like to leave feedback, here's my idea for how I'd like to implement it:

**Approach**

Since I really appreciate the clutter-free, easy-to-use interface, I want to keep the implementation simple: users can review either all of their flashcards or the flashcards of a single folder, subfolders included. On the UI, this would just be a "Review" button where the user picks whether cards should be answered with the term or the definition — everything else should ideally work automatically, e.g., tracking which cards are overdue and how many need to be reviewed.

I'm still experimenting with how to combine reviewing all flashcards vs. a single folder (with subfolders), but I plan to stick with a prominent "Review" section on "Home" and extend it with additional statistics (e.g. how many cards require review) later on. The "Review" button on top of a folder card will likely go away, since the folder itself is clickable and I noticed that nesting one clickable item inside another is confusing.

<img width="669" height="350" alt="Image" src="https://github.com/user-attachments/assets/acf1d322-6623-423e-b1a2-7c9c12391855" />

**Algorithm**

I'll go with FSRS v6 using the [ts-fsrs](https://github.com/open-spaced-repetition/ts-fsrs) library, with starting values of 6 and 24 hours for the learning steps and 4 hours for re-learning after a wrong answer on a previously learned card.

To keep it simple, I'm going with three answer options instead of four: "Don't know", "Took a while", and "Knew right away".

<img width="358" height="200" alt="Image" src="https://github.com/user-attachments/assets/63f880e5-ac35-4ca5-abf6-b937cf206086" />

Whenever "Don't know" is selected, the same card comes up again later in the same review session. Repeated answers on the same card within one session are not logged with the FSRS algorithm.

<img width="547" height="350" alt="Image" src="https://github.com/user-attachments/assets/627ea6d7-ec66-44f4-b0fd-d3cb7babaec7" />

**Current status**

I already have a basic implementation and am currently testing the algorithm with debug messages to see whether the behavior feels right in practice. (of course these will not be in the final release 🙂)

<img width="879" height="1024" alt="Image" src="https://github.com/user-attachments/assets/a774a689-a727-468f-8c88-c9438713e767" />

Happy to hear any thoughts! :)
